### ISSUE TITLE
Missing Data Types

### ISSUE THREAD

**Issue by amberream** [stoicflame/enunciate#495]
A few of my data types are not getting documented in the JSON or XML sections.  Here's the code for one of the missing types:

```
@JsonInclude(JsonInclude.Include.NON_NULL)
@XmlType(name = "DigitalID")
@XmlRootElement(name = "DigitalID")

public class DigitalIDModel
{
    public String keystorePassword;
    public String keyAlias;
    public String keyPassword;
    public String signReason;
    public String signLocation;
    public String signContactInformation;
    public boolean certifyingSignature;
    public int certifyingPermissions;

    public DigitalIDModel()
    {
        certifyingPermissions = 2;
    }
}
```

This class is referenced from the API in this method:

```
public SignatureFieldModel postSignature(@PathParam("fieldName") String fieldName,
        @FormDataParam("file") final InputStream file,
        @FormDataParam("digitalID") DigitalIDModel digitalID)
```

The others that are missing MAY be related to them being subclasses (this is the only thing I can think of that might be different from the others that ARE showing up).

```
@XmlType(name="NewPageAction")
@XmlRootElement(name="NewPageAction")
public class NewPageActionModel extends PageActionModel
{
    public int pageIndex = -1;
    public double width = 612;
    public double height = 792;
}
```

Here is the super class (this doesn't get documented and I wouldn't expect it to since it's abstract):

```
@JsonTypeInfo(use=JsonTypeInfo.Id.NAME, include=JsonTypeInfo.As.WRAPPER_OBJECT)
@XmlSeeAlso({OCRActionModel.class, ImageStampActionModel.class, TextStampActionModel.class, InsertAttachedPagesActionModel.class, InsertUploadedPagesActionModel.class, NewPageActionModel.class})
public abstract class PageActionModel
{

}
```

And the subclasses are only accessed in the API via polymorphism:

```
@POST
@Consumes(MediaType.MULTIPART_FORM_DATA)
public PagesModel post(@FormDataParam("pageAction") PageActionModel pageAction, @FormDataParam("file") InputStream file, @FormDataParam("file") FormDataContentDisposition fileDisposition)
```

Also, how do I get something to show up in the "description" column for the datatypes.  (I cut it out here, but I do have javadoc style comments on both the classes and variables).

**Comment by amberream** [comment]
I figured out how to include all my models: https://github.com/stoicflame/enunciate/wiki/Excluding-Including-Classes.

I am still not sure how to include the description of the DataType or the fields within it.

**Comment by stoicflame** [comment]
The code above doesn't have any description, so I wouldn't expect any to show up. Are you saying the real code has a description, but it's not showing up?

**Comment by amberream** [comment]
Yes, I cut out all the javadocs in my sample above, but here's a sample with docs... I don't see a description for any of my Models at Home/JSON or for any of the fields at Home/JSON/PageModel.

```
/**
 * A page in the document.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@XmlType(name = "Page")
@XmlRootElement(name = "Page")
public class PageModel
{
	/**
	 * The width of the page in pixels. Default value is 612.
	 */
	public double width = 612;

	/**
	 * The height of the page in pixels. Default value is 792.
	 */
	public double height = 792;

    ...
}
```

**Comment by stoicflame** [comment]
Does this help?

https://github.com/stoicflame/enunciate/wiki/Discovering-Source-Files

**Comment by amberream** [comment]
I solved the original problem of missing "Models" showing up under the JSON DataTypes with the sample shown on this page https://github.com/stoicflame/enunciate/wiki/Excluding-Including-Classes which seems to give the same example of using the "api-classes" element that the "Divcovering-Source-Files" link you shared above.  

The only problem I'm stuck with now is the missing descriptions.

**Comment by stoicflame** [comment]
Okay.

Well, if you've accounted for everything in that guide, I can't explain why you're not getting descriptions. In order for me to look deeper, I'm going to need some way to reproduce the issue (it works for me). Can you put together a sample project that exposes the problem?

**Comment by stoicflame** [comment]
Bump. Any updates on this?

**Comment by DrawGrumpy** [comment]
Hello,
I have some missing type too. Missing type is all a child in my hierarchy. 
but the parent class is present in my documentation API. I have put datatype-detection="aggressive" for jaxrs, jackson1 and jackson.

I have put <include pattern="com.gestion.facade.interfaces.ws.InstrumentChildJson" /> but the type InstrumentChildJson is not  in api.

the hierarchy is this : 
+ VehicleJson
 | + InstrumentJson
 |  | - InstrumentChildJson

in my api only VehicleJson is present. And all classes are in the same package.

**Comment by Thorndike** [comment]
I have tried to follow the listed instruction but with no luck. The only class I get into the documentation ist the abstract class that is in the API.
I would like some method to mark all subclasses to be included in the documentation and the preferably listed on the page of the abstract class with a link to the documentation of each class.

**Comment by stoicflame** [comment]
> I'd expect Enunciate to auto-add child classes (even when not annotated)

You mean there should be some logic that if `Pet` is added, then `Cat` and `Dog` should be automatically added?

That's actually harder than it sounds because it requires a double-pass through all the data types. The first pass would add the referenced classes, then you'd have to take a second pass to add all the subclasses of the referenced classes.

But I'll look at it closer if you want to open up a separate issue to track the work.

> I'd love to have a link to Cat and Dog in the Pet page (+1 for @Thorndike suggestion)

That's probably doable. Go ahead and open another issue to track the work.

> I'd love to be able to use one of the child classes in the documentation - for example a @documentationType doclet

See [Enunciate-Specific Annotations](https://github.com/stoicflame/enunciate/wiki/Enunciate-Specific-Annotations). Just use either `@TypeHint` or `@returnWrapped`.

**Comment by lbndev** [comment]
Hi,
Thanks for the answers.
I tested your suggestion with returnWrapped, but it doesn't match what I was thinking of.

I have this service :
`@Path("/PetStore")
public interface IPetStore {
    /**
     * Adopt a new pet (Cat or Dog)
     * @returnWrapped Cat
     */
    @GET @Path("/adopt")
    public Pet adopt();
}`

In the documentation, the return type of /PetStore/adopt is now shown as Cat. Which is wrong, because the real return type is Pet, and I could be getting a Dog instead of a cat.

I understand what returnWrapped is meant for (services returning a Response object), but my need is different here. I would like the return type to remain Pet, but still be able to show an example with a Cat implementation.

Ultimately we could have : 
`@Path("/PetStore")
public interface IPetStore {
    /**
     * Adopt a new pet (Cat or Dog)
     * @returnWrapped Pet
     * @documentationType Cat
     */
    @GET @Path("/adopt")
    public Response adopt();
}`

**Comment by stoicflame** [comment]
I see. Well, that doesn't seem too hard. Go ahead and open another issue for that, too.

**Comment by stoicflame** [comment]
Fix for #558 has been committed; I'm going to close this out for now.

If anyway can provide additional cases where types should be auto-detected, I'd be happy to re-open.

**Comment by stoicflame** [comment]
[Enunciate 2.8.0 has been released](https://github.com/stoicflame/enunciate/releases/tag/v2.8.0).
